Content.Server/Physics/Controllers/ChasingWalkSystem.cs

```diff
@@ -0,0 +1,95 @@
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+
+    private EntityQuery<TransformComponent> _xformQuery;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        _xformQuery = GetEntityQuery<TransformComponent>();
+    }
+
+    public override void UpdateBeforeSolve(bool prediction, float frameTime)
+    {
+        base.UpdateBeforeSolve(prediction, frameTime);
+
+        var query = EntityQueryEnumerator<ChasingWalkComponent>();
+        while (query.MoveNext(out var uid, out var chasing))
+        {
+            //Set Velocity to Target
+            if (chasing.NextImpulseTime <= _gameTiming.CurTime)
+            {
+                ForceImpulse(uid, chasing);
+                chasing.NextImpulseTime = _gameTiming.CurTime + TimeSpan.FromSeconds(chasing.ImpulseInterval);
+            }
+            //Change Target
+            if (chasing.NextChangeVectorTime <= _gameTiming.CurTime)
+            {
+                ChangeTarget(uid, chasing);
+
+                var delay = TimeSpan.FromSeconds(_random.NextFloat(chasing.ChangeVectorMinInterval, chasing.ChangeVectorMaxInterval));
+                chasing.NextChangeVectorTime = _gameTiming.CurTime + delay;
+            }
+        }
+    }
+
+    private void ChangeTarget(EntityUid uid, ChasingWalkComponent component)
+    {
+        //We find our coordinates and calculate the radius of the target search.
+        var xform = Transform(uid);
+        var range = component.MaxChaseRadius;
+        var compType = _random.Pick(component.ChasingComponent.Values).Component.GetType();
+        var allEnts = _lookup.GetEntitiesInRange(compType, _transform.GetMapCoordinates(xform), range);
+
+        //If there are no required components in the radius, don't moving.
+        if (allEnts.Count <= 0) return;
+
+        //In the case of finding required components, we choose a random one of them and remember its uid.
+        component.ChasingEntity = _random.Pick(allEnts).Owner;
+        component.Speed = _random.NextFloat(component.MinSpeed, component.MaxSpeed);
+    }
+
+    //pushing the entity toward its target
+    private void ForceImpulse(EntityUid uid, ChasingWalkComponent component)
+    {
+        if (Deleted(component.ChasingEntity) || component.ChasingEntity == null)
+        {
+            ChangeTarget(uid, component);
+            return;
+        }
+
+        if (!TryComp<PhysicsComponent>(uid, out var physics))
+            return;
+
+        //Calculating direction to the target.
+        var xform = Transform(component.ChasingEntity.Value);
+        var delta = _transform.GetMapCoordinates(xform).Position - _transform.GetMapCoordinates(_xformQuery.Get(uid)).Position;
+
```

Comment:
Resolving mapcoordinates is expensive and you should use InRange instead as it will avoid doing matrix transforms where possible.